core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java

```diff
@@ -219,8 +229,11 @@ public void close() {
 
   @Override
   public Stream<PageInfo> getPages() throws IOException {
-    Path rootDir = Paths.get(mRoot);
-    return Files.walk(rootDir).filter(Files::isRegularFile).map(this::getPageInfo);
+    Stream<Path> stream = Stream.empty();
+    for (Path root : mRoots) {
+      stream = Stream.concat(stream, Files.walk(root));
+    }
+    return stream.filter(Files::isRegularFile).map(this::getPageInfo);
```

Comment:
would it be more efficient if we could do filter and then do the concat?  I haven't look into the implementation or java's stream, so I'm just guessing

Reply:
will take a look

Reply:
```java
    /**
     * Creates a lazily concatenated stream whose elements are all the
     * elements of the first stream followed by all the elements of the
     * second stream.  The resulting stream is ordered if both
     * of the input streams are ordered, and parallel if either of the input
     * streams is parallel.  When the resulting stream is closed, the close
     * handlers for both input streams are invoked.
     *
     * @implNote
     * Use caution when constructing streams from repeated concatenation.
     * Accessing an element of a deeply concatenated stream can result in deep
     * call chains, or even {@code StackOverflowException}.
     *
     * @param <T> The type of stream elements
     * @param a the first stream
     * @param b the second stream
     * @return the concatenation of the two input streams
     */
    public static <T> Stream<T> concat(Stream<? extends T> a, Stream<? extends T> b) {
        Objects.requireNonNull(a);
        Objects.requireNonNull(b);

        @SuppressWarnings("unchecked")
        Spliterator<T> split = new Streams.ConcatSpliterator.OfRef<>(
                (Spliterator<T>) a.spliterator(), (Spliterator<T>) b.spliterator());
        Stream<T> stream = StreamSupport.stream(split, a.isParallel() || b.isParallel());
        return stream.onClose(Streams.composedClose(a, b));
    }
```

It seems no efficient down for `concat` operation.

Reply:
I see, looks like the concat is lazy, which should be just a O(1) operation.   Thank you Baolong for the clarification!